File: backend/routes/agents.py
"""Agent routes: list agents, ad-hoc chat with streaming."""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

import storage
from agents.registry import list_agents, create_runner, get_agent_config, save_agent_config, AGENT_NAMES

router = APIRouter()


class ChatRequest(BaseModel):
    message: str
    context: Optional[str] = None


class UpdateAgentConfigRequest(BaseModel):
    system_prompt: Optional[str] = None
    model: Optional[str] = None
    display_name: Optional[str] = None


@router.get("/projects/{project_id}/agents")
async def get_agents(project_id: str):
    project = storage.read_json(storage.project_json_path(project_id))
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return list_agents(project_id)


@router.get("/projects/{project_id}/agents/{agent_name}")
async def get_agent(project_id: str, agent_name: str):
    if agent_name not in AGENT_NAMES:
        raise HTTPException(status_code=404, detail=f"Unknown agent: {agent_name}")
    project = storage.read_json(storage.project_json_path(project_id))
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return get_agent_config(project_id, agent_name)


@router.put("/projects/{project_id}/agents/{agent_name}")
async def update_agent(project_id: str, agent_name: str, req: UpdateAgentConfigRequest):
    if agent_name not in AGENT_NAMES:
        raise HTTPException(status_code=404, detail=f"Unknown agent: {agent_name}")
    project = storage.read_json(storage.project_json_path(project_id))
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    overrides = req.model_dump(exclude_none=True)
    save_agent_config(project_id, agent_name, overrides)
    return get_agent_config(project_id, agent_name)


@router.post("/projects/{project_id}/agents/{agent_name}/chat")
async def chat_with_agent(project_id: str, agent_name: str, req: ChatRequest):
    if agent_name not in AGENT_NAMES:
        raise HTTPException(status_code=404, detail=f"Unknown agent: {agent_name}")
    project = storage.read_json(storage.project_json_path(project_id))
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")

    runner = create_runner(project_id, agent_name)

    async def generate():
        async for chunk in runner.stream(req.message, context=req.context):
            yield chunk

    return StreamingResponse(generate(), media_type="text/plain")

File: backend/src/routes/agents.ts
/**
 * Agent routes: list agents, get/update config, ad-hoc streaming chat.
 */

import * as storage from "../storage.js";
import { listAgents, getAgentConfig, saveAgentConfig, createRunner, AGENT_NAMES } from "../agents/registry.js";
import type { FastifyInstance } from "fastify";

export async function registerAgentRoutes(app: FastifyInstance): Promise<void> {
  // GET /projects/:projectId/agents
  app.get<{ Params: { projectId: string } }>("/projects/:projectId/agents", async (req, reply) => {
    const project = await storage.readJson(storage.projectJsonPath(req.params.projectId));
    if (!project) return reply.code(404).send({ detail: "Project not found" });
    return listAgents(req.params.projectId);
  });

  // GET /projects/:projectId/agents/:agentName
  app.get<{ Params: { projectId: string; agentName: string } }>(
    "/projects/:projectId/agents/:agentName",
    async (req, reply) => {
      if (!(AGENT_NAMES as readonly string[]).includes(req.params.agentName)) {
        return reply.code(404).send({ detail: `Unknown agent: ${req.params.agentName}` });
      }
      const project = await storage.readJson(storage.projectJsonPath(req.params.projectId));
      if (!project) return reply.code(404).send({ detail: "Project not found" });
      return getAgentConfig(req.params.projectId, req.params.agentName);
    },
  );

  // PUT /projects/:projectId/agents/:agentName
  app.put<{
    Params: { projectId: string; agentName: string };
    Body: { system_prompt?: string; model?: string; display_name?: string };
  }>("/projects/:projectId/agents/:agentName", async (req, reply) => {
    if (!(AGENT_NAMES as readonly string[]).includes(req.params.agentName)) {
      return reply.code(404).send({ detail: `Unknown agent: ${req.params.agentName}` });
    }
    const project = await storage.readJson(storage.projectJsonPath(req.params.projectId));
    if (!project) return reply.code(404).send({ detail: "Project not found" });

    const overrides: Record<string, unknown> = {};
    if (req.body.system_prompt !== undefined) overrides.system_prompt = req.body.system_prompt;
    if (req.body.model !== undefined) overrides.model = req.body.model;
    if (req.body.display_name !== undefined) overrides.display_name = req.body.display_name;

    await saveAgentConfig(req.params.projectId, req.params.agentName, overrides);
    return getAgentConfig(req.params.projectId, req.params.agentName);
  });

  // POST /projects/:projectId/agents/:agentName/chat — streaming text/plain
  app.post<{
    Params: { projectId: string; agentName: string };
    Body: { message: string; context?: string };
  }>("/projects/:projectId/agents/:agentName/chat", async (req, reply) => {
    if (!(AGENT_NAMES as readonly string[]).includes(req.params.agentName)) {
      return reply.code(404).send({ detail: `Unknown agent: ${req.params.agentName}` });
    }
    const project = await storage.readJson(storage.projectJsonPath(req.params.projectId));
    if (!project) return reply.code(404).send({ detail: "Project not found" });

    const runner = await createRunner(req.params.projectId, req.params.agentName);

    reply.raw.writeHead(200, {
      "Content-Type": "text/plain",
      "Transfer-Encoding": "chunked",
      "Cache-Control": "no-cache",
    });

    try {
      for await (const chunk of runner.stream(req.body.message, req.body.context)) {
        reply.raw.write(chunk);
      }
    } finally {
      reply.raw.end();
    }

    return reply;
  });
}
